### PR TITLE
PHP: php server commit 4/n, add RpcServer, generator and server examples

### PR DESCRIPTION
**Pull request by HannahShiSFB**
add RpcServer and generator.
add server examples: greeter and route guide.
update readme

@markdroth
@stanley-cheung

### CODE REVIEW

File: examples/php/greeter_and_routeguide_multi_server.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+require dirname(__FILE__) . '/../../src/php/lib/Grpc/MethodDescriptor.php';
+require dirname(__FILE__) . '/../../src/php/lib/Grpc/Status.php';
+require dirname(__FILE__) . '/../../src/php/lib/Grpc/ServerCallReader.php';
+require dirname(__FILE__) . '/../../src/php/lib/Grpc/ServerCallWriter.php';
+require dirname(__FILE__) . '/../../src/php/lib/Grpc/ServerContext.php';
+require dirname(__FILE__) . '/../../src/php/lib/Grpc/RpcServer.php';
+require dirname(__FILE__) . '/vendor/autoload.php';
+require dirname(__FILE__) . '/route_guide/RouteGuideService.php';
+
+class Greeter extends \Helloworld\GreeterServiceStub
```
Comment (stanley-cheung) [inline examples/php/greeter_and_routeguide_multi_server.php:29]: Where is the GreeterServiceStub checked in?

File: examples/php/route_guide/RouteGuideService.php [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+
+class RouteGuideService extends \Routeguide\RouteGuideServiceStub
```
Comment (stanley-cheung) [inline examples/php/route_guide/RouteGuideService.php:21]: Same here. Where is the RouteGuideServiceStub checked in?

File: examples/php/greeter_and_routeguide_multi_server.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+require dirname(__FILE__) . '/../../src/php/lib/Grpc/MethodDescriptor.php';
```
Comment (stanley-cheung) [inline examples/php/greeter_and_routeguide_multi_server.php:20]: Do we need the following 6 lines? I thought those will be available through `vendor/autoload.php`

File: examples/php/greeter_server.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+require dirname(__FILE__) . '/../../src/php/lib/Grpc/MethodDescriptor.php';
```
Comment (stanley-cheung) [inline examples/php/greeter_server.php:20]: Same here. I think these should be available through vendor/autoload.php

File: examples/php/greeter_and_routeguide_multi_server.php (hunk 1 above)
Comment (HannahShiSFB) [reply examples/php/greeter_and_routeguide_multi_server.php:29]: added

File: examples/php/route_guide/RouteGuideService.php (hunk 2 above)
Comment (HannahShiSFB) [reply examples/php/route_guide/RouteGuideService.php:21]: added

File: examples/php/greeter_server.php (hunk 4 above)
Comment (HannahShiSFB) [reply examples/php/greeter_server.php:20]: It's temporary. These lines will be deleted after php server feature release.

File: examples/php/greeter_server.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+class Greeter extends Helloworld\GreeterServiceStub
+{
+    public function SayHello(
+        \Helloworld\HelloRequest $request,
+        \Grpc\ServerContext $serverContext
+    ): array {
+        $name = $request->getName();
+        $response = new \Helloworld\HelloReply();
+        $response->setMessage("Hello " . $name);
+        return [$response];
```
Comment (stanley-cheung) [inline examples/php/greeter_server.php:37]: Why does this `[$response]` need to be an array? Shouldn't it just be `return $response;`? Same for all the unary methods in the rest of the examples.

I see. So user is expecting to return `list($response, $initialMetadata, $status)` here? Then let's figure out what's a better way to do this.

As a user I expect to write `return $response;`. If I need to return initial metadata and status, perhaps we should do it via the `$serverContext`. Let's see how other languages do this.

Yea, check out how Ruby implemented the interop test [here](https://github.com/grpc/grpc/blob/master/src/ruby/pb/test/server.rb#L179-L185). The return value of the RPC method itself is just the response proto. Everything else, like the status code, message, metadata are all set to the `_call` object. For us here in PHP, we can do it via this `$serverContext` object.

File: examples/php/route_guide/RouteGuideService.php [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        $notFoundFeature = new Routeguide\Feature([
+            'name' => '',
+            'location' => $request,
+        ]);
+        return [$feature ?? $notFoundFeature];
+    }
+
+    public function ListFeatures(
+        \Routeguide\Rectangle $request,
+        \Grpc\ServerCallWriter $writter,
```
Comment (stanley-cheung) [inline examples/php/route_guide/RouteGuideService.php:104]: Typo, `$writer`. It's also in the code generator and generated code too.

Comment (HannahShiSFB) [reply examples/php/route_guide/RouteGuideService.php:104]: fixed

File: examples/php/greeter_server.php (hunk 5 above)
Comment (HannahShiSFB) [reply examples/php/greeter_server.php:37]: conform to the existing greeter client.
[list($response, $status) = $client->SayHello($request)->wait();](https://github.com/grpc/grpc/blob/b516dbfa21c5a2065f8a99a5890316d7da75d238/examples/php/greeter_client.php#L34)

File: examples/php/greeter_and_routeguide_multi_server.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+class Greeter extends \Helloworld\GreeterServiceStub
+{
+    public function SayHello(
+        \Helloworld\HelloRequest $request,
+        \Grpc\ServerContext $serverContext
+    ): array {
+        $name = $request->getName();
+        $response = new \Helloworld\HelloReply();
+        $response->setMessage("Hello " . $name);
+        return [$response];
```
Comment (stanley-cheung) [inline examples/php/greeter_and_routeguide_multi_server.php:38]: Missed changing this instance?

**Review by stanley-cheung — APPROVED** [review]

File: examples/php/greeter_and_routeguide_multi_server.php (hunk 7 above)
Comment (HannahShiSFB) [reply examples/php/greeter_and_routeguide_multi_server.php]: sorry...missed 3

**Comment by stanley-cheung** [comment]
I patched in this PR to my xDS test workspace and I got this error:

```
Fatal error: Uncaught Error: Call to undefined method LoadBalancerStatsService::getMethodDescriptors() in /ho\
me/stanleycheung/grpc/src/php/lib/Grpc/RpcServer.php:50                                                       
Stack trace:                                                                                                  
#0 /home/stanleycheung/grpc/src/php/tests/interop/xds_client.php(504): Grpc\RpcServer->handle(Object(LoadBala\
ncerStatsService))                                                                                            
#1 {main}                                                                                                     
  thrown in /home/stanleycheung/grpc/src/php/lib/Grpc/RpcServer.php on line 50   
```

It's likely we need to re-run `src/php/bin/generate_proto_php.sh` to re-generate those generated files after we update the protoc plugin?

File: src/compiler/php_generator_helpers.h [hunk 8]
```diff
@@ -28,8 +28,11 @@ namespace grpc_php_generator {
 
 inline std::string GetPHPServiceClassname(
     const grpc::protobuf::ServiceDescriptor* service,
-    const std::string& class_suffix) {
-  return service->name() + (class_suffix == "" ? "Client" : class_suffix);
+    const std::string& class_suffix, bool is_server) {
+  return service->name() +
+         (class_suffix == "" ? (is_server ? "Service" : "Client")
```
Comment (stanley-cheung) [inline src/compiler/php_generator_helpers.h:33]: The dilemma here is that, if the service is already named `TestService` for example, then we will be generating the class `TestServiceServiceStub` in `TestServiceServiceStub.php`.

Python would have called it `TestServiceServicer`. C# calls it `TestServiceBase`. Ruby and C++ just call it `TestService`.

I am not sure what the best name should be...

**Comment by stanley-cheung** [comment]
PHP xDS v3 tests all passed with changes in this PR: https://fusion2.corp.google.com/invocations/15697ebe-23e3-445b-b99f-35b3b7f3919f/targets

PR tests are passing.

Reviewed code and changes.

**Pull request merged**